core/src/main/java/com/linecorp/armeria/common/util/SystemInfo.java

```diff
@@ -377,4 +394,74 @@ private static void logFailure(String method, boolean warn, Throwable cause) {
             }
         }
     }
+
+    private static final class DefaultNonLoopbackIPv4Address {
+
+        // Forked from InetUtils in spring-cloud-common 3.0.0.M1 at e7bb7ed3ae19a91c6fa7b3b698dd9788f70df7d4
+        // - Use CIDR in isPreferredAddress instead of regular expression.
+
+        @Nullable
+        static final Inet4Address defaultNonLoopbackIpV4Address;
+
+        static {
+            Inet4Address result = null;
+            try {
+                int lowest = Integer.MAX_VALUE;
+                for (final Enumeration<NetworkInterface> nics = NetworkInterface.getNetworkInterfaces();
+                     nics.hasMoreElements();) {
+                    final NetworkInterface ifc = nics.nextElement();
+                    if (ifc.isUp()) {
+                        // The NIC whose index is the lowest will be likely the valid IPv4 address.
+                        // See https://github.com/spring-cloud/spring-cloud-commons/issues/82.
+                        if (ifc.getIndex() < lowest || result == null) {
+                            lowest = ifc.getIndex();
+                        } else if (result != null) {
+                            continue;
+                        }
+
+                        for (final Enumeration<InetAddress> addrs = ifc.getInetAddresses();
+                             addrs.hasMoreElements();) {
+                            final InetAddress address = addrs.nextElement();
+                            if (address instanceof Inet4Address &&
+                                !address.isLoopbackAddress() &&
+                                isPreferredAddress(address)) {
+                                result = (Inet4Address) address;
+                            }
+                        }
+                    }
+                }
+            } catch (IOException ex) {
+                logger.error("Cannot get non-loopback IPv4 address.", ex);
```

Comment:
- Cannot -> Could not ?
- non-loopback -> a non-lookback
- . -> :
- Log level 'error' means the situation that an application cannot work anymore due to an unrecoverable error. It seems not?